lib/host/wasi/inode-win.cpp

```diff
@@ -82,53 +434,310 @@ INode INode::stdErr() noexcept {
+  // bytes with zeros
+
+  FILE_STANDARD_INFO StandardInfo;
+  FILE_ALLOCATION_INFO AllocationInfo;
+
+  if (unlikely(GetFileInformationByHandleEx(Handle, FileStandardInfo,
+                                            &StandardInfo,
+                                            sizeof(StandardInfo))) == FALSE) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  // Update the size attribute
+  AllocationInfo.AllocationSize = toLargeIntegerFromUnsigned(Size);
+
+  if (SetFileInformationByHandle(Handle, FileAllocationInfo, &AllocationInfo,
+                                 sizeof(AllocationInfo)) == FALSE) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  return {};
 }
 
-WasiExpect<void> INode::fdFilestatSetTimes(__wasi_timestamp_t,
-                                           __wasi_timestamp_t,
-                                           __wasi_fstflags_t) const noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+WasiExpect<void>
+INode::fdFilestatSetTimes(__wasi_timestamp_t ATim, __wasi_timestamp_t MTim,
+                          __wasi_fstflags_t FstFlags) const noexcept {
+
+  // Let FileTime be initialized to zero if the times need not be changed
+  FILETIME AFileTime = {0, 0};
+  FILETIME MFileTime = {0, 0};
+
+  // For setting access time
+  if (FstFlags & __WASI_FSTFLAGS_ATIM) {
+    uint64_t Aticks = ATim / NANOSECONDS_PER_TICK + TICKS_TO_UNIX_EPOCH;
+    AFileTime.dwLowDateTime = static_cast<DWORD>(Aticks % 0x100000000ULL);
+    AFileTime.dwHighDateTime = static_cast<DWORD>(Aticks / 0x100000000ULL);
+  } else if (FstFlags & __WASI_FSTFLAGS_ATIM_NOW) {
+    GetSystemTimeAsFileTime(&AFileTime);
+  }
+
+  // For setting modification time
+  if (FstFlags & __WASI_FSTFLAGS_MTIM) {
+    uint64_t Mticks = MTim / NANOSECONDS_PER_TICK + TICKS_TO_UNIX_EPOCH;
+    MFileTime.dwLowDateTime = static_cast<DWORD>(Mticks % 0x100000000ULL);
+    MFileTime.dwHighDateTime = static_cast<DWORD>(Mticks / 0x100000000ULL);
+  } else if (FstFlags & __WASI_FSTFLAGS_MTIM_NOW) {
+    GetSystemTimeAsFileTime(&MFileTime);
+  }
+
+  if (unlikely(SetFileTime(Handle, nullptr, &AFileTime, &MFileTime)) == FALSE) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  return {};
 }
 
-WasiExpect<void> INode::fdPread(Span<Span<uint8_t>>, __wasi_filesize_t,
-                                __wasi_size_t &) const noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+WasiExpect<void> INode::fdPread(Span<Span<uint8_t>> IOVs,
+                                __wasi_filesize_t Offset,
+                                __wasi_size_t &NRead) const noexcept {
+  NRead = 0;
+  uint64_t LocalOffset = Offset;
+
+  for (auto IOV : IOVs) {
+    DWORD NumberOfBytesRead = 0;
+    OVERLAPPED Result;
+
+    Result.Offset = static_cast<uint32_t>(LocalOffset);
+    Result.OffsetHigh = static_cast<uint32_t>(LocalOffset >> 32);
+
+    // There maybe issues due to casting IOV.size() to unit32_t
```

Comment:
What's the issue?

Reply:
`IOV.size()` returns a 64 bit integer but the underling windows API uses a 32 bit integer. So can there be any issues due to the narrowing of the range?